pkg/sidecar_controller/snapshot_controller.go

```diff
@@ -0,0 +1,576 @@
+		return nil
+	}
+	contentClone := content.DeepCopy()
+	statusError := &crdv1.VolumeSnapshotError{
+		Time: &metav1.Time{
+			Time: time.Now(),
+		},
+		Message: &message,
+	}
+	contentClone.Status.Error = statusError
+	ready := false
+	contentClone.Status.ReadyToUse = &ready
+	newContent, err := ctrl.clientset.SnapshotV1beta1().VolumeSnapshotContents().UpdateStatus(contentClone)
+
+	if err != nil {
+		klog.V(4).Infof("updating VolumeSnapshotContent[%s] error status failed %v", uniqueKey, err)
+		return err
+	}
+
+	_, err = ctrl.storeContentUpdate(newContent)
+	if err != nil {
+		klog.V(4).Infof("updating VolumeSnapshotContent[%s] error status: cannot update internal cache %v", uniqueKey, err)
+		return err
+	}
+	// Emit the event only when the status change happens
+	ctrl.eventRecorder.Event(newContent, eventtype, reason, message)
+
+	return nil
+}
+
+func (ctrl *csiSnapshotSideCarController) getCSISnapshotInput(content *crdv1.VolumeSnapshotContent) (*crdv1.VolumeSnapshotClass, map[string]string, error) {
+	uniqueKey := utils.SnapshotOrContentKey(content)
+	className := content.Spec.VolumeSnapshotClassName
+	klog.V(5).Infof("getCSISnapshotInput for content [%s]: VolumeSnapshotClassName [%s]", content.Name, *className)
+	var class *crdv1.VolumeSnapshotClass
+	var err error
+	if className != nil {
+		class, err = ctrl.getSnapshotClass(*className)
+		if err != nil {
+			klog.Errorf("getCSISnapshotInput failed to getClassFromVolumeSnapshot %s", err)
+			return nil, nil, err
+		}
+	} else {
+		klog.Errorf("failed to getCSISnapshotInput %s without a snapshot class", uniqueKey)
+		return nil, nil, fmt.Errorf("failed to take snapshot %s without a snapshot class", uniqueKey)
+	}
+
+	// Resolve snapshotting secret credentials.
+	snapshotterCredentials, err := ctrl.GetCredentialsFromAnnotation(content)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return class, snapshotterCredentials, nil
+}
+
+func (ctrl *csiSnapshotSideCarController) checkandUpdateBoundSnapshotStatusOperation(content *crdv1.VolumeSnapshotContent) (*crdv1.VolumeSnapshotContent, error) {
+	var err error
+	var creationTime time.Time
+	var size int64
+	var readyToUse = false
+	var driverName string
+	var snapshotID string
+
+	if content.Spec.Source.VolumeHandle == nil && content.Spec.Source.SnapshotHandle != nil {
+		klog.V(5).Infof("checkandUpdateBoundSnapshotStatusOperation: checking whether snapshot is pre-bound to content [%s]", content.Name)
+		readyToUse, creationTime, size, err = ctrl.handler.GetSnapshotStatus(content)
+		if err != nil {
+			klog.Errorf("checkandUpdateBoundSnapshotStatusOperation: failed to call get snapshot status to check whether snapshot is ready to use %q", err)
+			return nil, err
+		}
+		driverName = content.Spec.Driver
+		if content.Spec.Source.SnapshotHandle != nil {
+			snapshotID = *content.Spec.Source.SnapshotHandle
+		}
+	} else if content.Spec.Source.VolumeHandle != nil {
```

Comment:
If content.Status.SnapshotHandle already exists, we should just call ListSnapshots?